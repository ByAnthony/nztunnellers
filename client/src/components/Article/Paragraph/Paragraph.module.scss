@use 'src/index.scss';

.paragraph {
  margin: 0 auto;

  p {
    font-size: 1.15rem;
    font-weight: 200;
    line-height: 1.8rem;

    @include index.desktop {
      font-size: 1.25rem;
      line-height: 2.25rem;
    }
  }

  a {
    font-size: 0.9rem;
    vertical-align: super;
    line-height: 80%;
    color: index.$secondary-color;
  }

  @include index.desktop {
    margin: 0 auto;
    width: 42rem;
  }
}
