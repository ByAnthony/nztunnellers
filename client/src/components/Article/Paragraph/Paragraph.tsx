import { formatText } from '../../../utils/text-utils';

import { Section } from '../../../types/article';

import STYLES from './Paragraph.module.scss';

type Props = {
    section: Section;
}

export function Paragraph({ section }: Props) {
  return (
    <div className={STYLES.paragraph}>
      <h2>{section.title}</h2>
      {formatText(section.text)}
    </div>
  );
}
