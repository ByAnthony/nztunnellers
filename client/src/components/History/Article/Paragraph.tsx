import { Section } from '../../../types/article';
import STYLES from '../History.module.scss';

type Props = {
    section: Section;
}

export const formatText = (text: string) => {
  const paragraphs = text.split('\\n\\n');
  return paragraphs.map((paragraph) => (
    <p key={paragraphs.indexOf(paragraph)}>
      {paragraph
        .replace(/--/g, '\u00A0')
        .split(/(\*.+?\*)|(\[.+?\))/g)
        .map((segment) => {
          if (segment && segment.startsWith('*') && segment.endsWith('*')) {
            const italicText = segment.slice(1, -1);
            return <em key={paragraph.indexOf(segment)}>{italicText}</em>;
          }
          if (segment && segment.startsWith('[') && segment.endsWith(')')) {
            const linkText = segment.slice(1, -1);
            const [label, url] = linkText.split('](');
            if (url.includes('footnote')) {
              return (
                <a key={paragraph.indexOf(segment)} href={url} id={`reference_${label}`}>
                  [
                  {label}
                  ]
                </a>
              );
            }
            const number = label.slice(0, -1);
            return (
              <a key={paragraph.indexOf(segment)} href={url} id={`footnote_${number}`}>
                {label}
              </a>
            );
          }
          return <span key={paragraph.indexOf(segment)}>{segment}</span>;
        })}
    </p>
  ));
};

export function Paragraph({ section }: Props) {
  return (
    <div className={STYLES.paragraph}>
      <h2>{section.title}</h2>
      {formatText(section.text)}
    </div>
  );
}
