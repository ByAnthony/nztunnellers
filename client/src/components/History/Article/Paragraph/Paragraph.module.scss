@use 'src/index.scss';

.paragraph {
  margin: 0 auto;

  p {
    font-size: 1.15rem;
    font-weight: 300;
    line-height: 1.8rem;

    @include index.desktop {
      font-size: 1.25rem;
      line-height: 2.1rem;
    }

    @include index.large-desktop {
      font-size: 1.5rem;
      line-height: 2.5rem;
    }
  }

  a {
    font-size: 0.9rem;
    vertical-align: super;
    line-height: 80%;
    color: index.$secondary-color;
  }

  @include index.desktop {
    @media screen and (min-width: 62rem) and (max-width: 106rem) {
      margin: 0 auto;
      width: 65%;
    }
  }

  @include index.large-desktop {
    margin: 0 auto;
    width: 65%;
  }
}
