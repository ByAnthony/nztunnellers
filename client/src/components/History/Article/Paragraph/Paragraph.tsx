import { Section } from '../../../../types/article';
import STYLES from './Paragraph.module.scss';

type Props = {
    section: Section;
}

export const formatText = (text: string) => {
  const paragraphs = text.split('\\n\\n');
  let segmentKey: number = 0;
  return paragraphs.map((paragraph) => {
    const segments = paragraph
      .replace(/--/g, '\u00A0')
      .split(/(\*.+?\*)|(\[.+?\))/g);
    const formattedSegments = segments.map((segment) => {
      segmentKey += 1;
      if (segment && segment.startsWith('*') && segment.endsWith('*')) {
        const italicText = segment.slice(1, -1);
        return <em key={segmentKey}>{italicText}</em>;
      }
      if (segment && segment.startsWith('[') && segment.endsWith(')')) {
        const linkText = segment.slice(1, -1);
        const [label, url] = linkText.split('](');
        if (url.includes('footnote')) {
          return (
            <a key={segmentKey} href={url} id={`reference_${label}`}>
              [
              {label}
              ]
            </a>
          );
        }
        const number = label.slice(0, -1);
        return (
          <a key={segmentKey} href={url} id={`footnote_${number}`}>
            {label}
          </a>
        );
      }
      return <span key={segmentKey}>{segment}</span>;
    });
    return <p key={paragraphs.indexOf(paragraph)}>{formattedSegments}</p>;
  });
};

export function Paragraph({ section }: Props) {
  return (
    <div className={STYLES.paragraph}>
      <h2>{section.title}</h2>
      {formatText(section.text)}
    </div>
  );
}
