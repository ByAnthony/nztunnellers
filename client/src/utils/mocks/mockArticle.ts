import { AboutUs } from '../../types/about-us';
import {
  Article, Image, Next, Section,
} from '../../types/article';

export const mockTitle: string = 'My Awesome\\Article Title';

export const mockChapter: number = 1;

export const mockTopImage: Image = {
  file: 'img-123.png',
  alt: 'Accessible alt text',
};

export const mockArticleImage: Image = {
  ...mockTopImage,
  title: 'An Amazing Photo',
  photographer: 'The French Photographer Doisneau',
  reference: 'Where the photo is preserved',
};

export const mockSectionOne: Section = {
  title: 'Section Title',
  text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea *commodo consequat*. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\\n\\nLorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat[1](#footnote_1). Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.\\n\\n(6--July 2023)',
};

export const mockSectionTwo: Section = {
  title: 'Section Title',
  text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea *commodo consequat*. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\\n\\nLorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat[1](#footnote_1). Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.\\n\\n(6--July 2023)',
};

export const mockNextButton: Next = {
  url: 'my-path-to-next-chapter/',
  title: 'Next\\Chapter',
  chapter: 3,
};

export const mockNotes: string = '[1.](#reference_1) Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\\n\\n[2.](#reference_2) Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';

export const mockArticle: Article = {
  id: 'my-awesome-article',
  chapter: mockChapter,
  title: mockTitle,
  section: [mockSectionOne, mockSectionTwo],
  image: [mockTopImage, mockArticleImage],
  next: mockNextButton,
  notes: mockNotes,
};

export const mockAboutUs: AboutUs = {
  id: 'my-awesome-article',
  title: mockTitle,
  section: [mockSectionOne, mockSectionTwo],
  image: [mockTopImage],
};
