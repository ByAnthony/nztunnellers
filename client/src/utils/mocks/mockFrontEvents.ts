import { Date, Event, EventDetail } from '../../types/tunneller';

const mockDate = (year: string | null, dayMonth: string | null): Date => ({
  year,
  dayMonth,
});

export const mockEvent: EventDetail = {
  description: null,
  title: null,
  image: null,
};

export const mockEventCompany: EventDetail = {
  ...mockEvent,
  description: 'Something happened that day for the Company',
  title: 'The Company',
  image: 'image.jpg',
};
export const mockEventDescription: EventDetail = { ...mockEvent, description: 'Something happened that day' };
export const mockEventEnlisted: EventDetail = { ...mockEvent, description: 'Main Body', title: 'Enlisted' };
export const mockEventPosted: EventDetail = { ...mockEvent, description: 'Reinforcement', title: 'Posted' };
export const mockEventTrained: EventDetail = { ...mockEvent, description: 'Military Camp, Somewhere', title: 'Trained' };
export const mockEventTitleAndDescription: EventDetail = { ...mockEvent, description: 'Something happened', title: 'Major Event' };

export const mockEventDiedOfDisease: EventDetail = { ...mockEvent, description: 'Military Hospital', title: 'Died of disease' };
export const mockEventKilledInAction: EventDetail = { ...mockEvent, description: 'Killed on the battlefield', title: 'Killed in action' };
export const mockEventDiedOfWounds: EventDetail = { ...mockEvent, description: 'Wounded on the battlefield', title: 'Died of wounds' };
export const mockEventDiedOfAccident: EventDetail = { ...mockEvent, description: 'Fatal accident', title: 'Died of accident' };
export const mockEventBuried: EventDetail = { ...mockEvent, description: 'Cemetery, France', title: 'Buried' };
export const mockEventGrave: EventDetail = { ...mockEvent, description: '1 A 26', title: 'Grave reference' };

const mockDeathEvent = (deathEvent: EventDetail): Record<string, Event[]> => ({
  1916: [
    {
      date: mockDate('1916', '27 September'),
      event: [mockEventCompany],
    },
    {
      date: mockDate('1916', '19 November'),
      event: [mockEventDescription],
    },
  ],
  1917: [
    {
      date: mockDate('1917', '8 August'),
      event: [mockEventDescription],
    },
    {
      date: mockDate('1917', '8 August'),
      event: [deathEvent, mockEventBuried, mockEventGrave],
    },
  ],
});

export const mockFrontEventsWithDiedOfDisease:
    Record<string, Event[]> = mockDeathEvent(mockEventDiedOfDisease);

export const mockFrontEventsWithKilledInAction:
    Record<string, Event[]> = mockDeathEvent(mockEventKilledInAction);

export const mockFrontEventsWithDiedOfWounds:
    Record<string, Event[]> = mockDeathEvent(mockEventDiedOfWounds);

export const mockFrontEventsWithDiedOfAccident:
    Record<string, Event[]> = mockDeathEvent(mockEventDiedOfAccident);

export const mockFrontEvents: Record<string, Event[]> = {
  1915: [
    {
      date: mockDate('1915', '26 January'),
      event: [mockEventEnlisted],
    },
    {
      date: mockDate('1915', '30 January'),
      event: [mockEventPosted],
    },

    {
      date: mockDate('1915', '14 February'),
      event: [mockEventTrained],
    },
    {
      date: mockDate('1915', '4 May'),
      event: [mockEventTitleAndDescription],
    },
  ],
  1916: [
    {
      date: mockDate('1916', '23 June'),
      event: [mockEventDescription, mockEventDescription],
    },
    {
      date: mockDate('1916', '27 September'),
      event: [mockEventCompany],
    },
    {
      date: mockDate('1916', '19 November'),
      event: [mockEventDescription],
    },
    {
      date: mockDate('1916', '12 December'),
      event: [mockEventCompany, mockEventDescription],
    },
  ],
  1917: [
    {
      date: mockDate('1917', '9 April'),
      event: [mockEventTitleAndDescription, mockEventCompany],
    },
    {
      date: mockDate('1917', '8 August'),
      event: [mockEventDescription],
    },
  ],
  1918: [
    {
      date: mockDate('1918', '3 March April'),
      event: [mockEventTitleAndDescription],
    },
    {
      date: mockDate('1918', '13 October'),
      event: [mockEventDescription],
    },
  ],
  1919: [
    {
      date: mockDate('1919', '4 April'),
      event: [mockEventTitleAndDescription],
    },
    {
      date: mockDate('1919', '12 July'),
      event: [mockEventDescription],
    },
  ],
};
