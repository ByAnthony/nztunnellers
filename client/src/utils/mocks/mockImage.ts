import {
  Image, ImageSource, ImageNewspaper, ImageArchives, ImageBook,
} from '../../types/tunneller';

export const mockImageAucklandLibraries: string = 'https://digitalnz.org/records?text=31-B2671&tab=Images#';

export const mockImageArchives: ImageArchives = {
  location: 'Auckland War Memorial Museum',
  reference: 'MS-93/157',
};

export const mockImageFamily: string = 'Courtesy of John Doe family';

export const mockImageNewspaper: ImageNewspaper = {
  name: 'Auckland Star',
  date: '12 July 1898',
};

export const mockImageBook: ImageBook = {
  title: 'My Mock Title',
  town: 'Bethune',
  publisher: 'Publisher and Co.',
  authors: [
    {
      forename: 'Jane',
      surname: 'Doe',
    },
  ],
  year: '1925',
  page: 'p. 89',
};

export const mockImageSource: ImageSource = {
  aucklandLibraries: null,
  archives: null,
  family: null,
  newspaper: null,
  book: null,
};

export const mockImage: Image = {
  url: '1/1234.jpg',
  source: mockImageSource,
};
