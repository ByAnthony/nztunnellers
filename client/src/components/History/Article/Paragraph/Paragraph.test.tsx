import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { mockSection } from '../../../../utils/mocks/mockArticle';
import { Paragraph } from './Paragraph';

const component = (
  <Paragraph section={mockSection} />
);

test('renders the component correctly', () => {
  const { asFragment } = render(component);

  expect(asFragment()).toMatchSnapshot();
});

test('renders title correctly', () => {
  render(component);

  expect(screen.getByText(/Section Title/)).toBeInTheDocument();
  expect(screen.getAllByRole('link')[0]).toHaveAttribute('href', '#footnote_1');
  expect(screen.getAllByRole('link')[0]).toHaveAttribute('id', 'reference_1');
  expect(screen.getAllByRole('link')[1]).toHaveAttribute('href', '#reference_1');
  expect(screen.getAllByRole('link')[1]).toHaveAttribute('id', 'footnote_1');
});
