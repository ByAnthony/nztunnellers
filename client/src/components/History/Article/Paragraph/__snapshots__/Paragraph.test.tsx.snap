// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component correctly 1`] = `
<DocumentFragment>
  <div
    class="paragraph"
  >
    <h2>
      Section Title
    </h2>
    <p>
      <span>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea 
      </span>
      <em>
        commodo consequat
      </em>
      <span />
      <span>
        . Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
      </span>
    </p>
    <p>
      <span>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat
      </span>
      <span />
      <a
        href="#footnote_1"
        id="reference_1"
      >
        [1.]
      </a>
      <span>
        . Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
      </span>
    </p>
    <p>
      <span>
        (6Â July 2023)
      </span>
    </p>
    <p>
      <span />
      <span />
      <a
        href="#reference_1"
        id="footnote_1"
      >
        1.
      </a>
      <span>
         Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
      </span>
    </p>
  </div>
</DocumentFragment>
`;
